qutip/tests/test_interpolate.py

```diff
@@ -30,325 +30,123 @@
-    assert_(np.max(np.abs(out1-out2)) < 1e-4)
-
-
-@pytest.mark.slow
-@unittest.skipIf(not Cython_OK, 'Cython not found or version too low.')
-def test_interpolate_brevolve2():
-    """
-    Interpolate: brmesolve str + interp (complex)
-    """
-    tlist = np.linspace(0,5,50)
-    y = 0.1j*np.sin(tlist)
-    S = Cubic_Spline(tlist[0], tlist[-1], y)
-    N = 10
-    psi0 = fock_dm(N,1)
-    a = destroy(N)
-    H = [a.dag()*a,[a**2+a.dag()**2,'0.1j*sin(t)'], [a**2+a.dag()**2,'-0.1j*sin(t)']]
-    H2 = [a.dag()*a, [a**2+a.dag()**2, S], [a**2+a.dag()**2,'-0.1j*sin(t)']]
-    out1 = mesolve(H, psi0, tlist, [], [a.dag()*a]).expect[0]
-    out2 = brmesolve(H2, psi0, tlist, a_ops=[], e_ops=[a.dag()*a]).expect[0]
-    assert_(np.max(np.abs(out1-out2)) < 1e-4)
+import scipy.interpolate
+import pytest
+import qutip
+
+pytestmark = [pytest.mark.usefixtures("in_temporary_directory")]
+
+
+def pytest_generate_tests(metafunc):
+    """
+    Perform more complex parametrisation logic for tests in this module.  This
+    was originally needed because the `brmesolve` solver cannot mix splines and
+    functions in the Hamiltonian (it can only accept string-formatted
+    time-dependence), but we want to parametrize this for all the other tests.
+    """
+    if ("solver" in metafunc.fixturenames
+            and "coefficients" in metafunc.fixturenames):
+        _parametrize_solver_coefficients(metafunc)
+
+
+@pytest.mark.parametrize('noise', [
+    pytest.param(lambda n: 0.1*np.random.rand(n), id="real"),
+    pytest.param(lambda n: 0.1j*np.random.rand(n), id="complex"),
+])
+def test_equivalence_to_scipy(noise):
+    x = np.linspace(0, 2*np.pi, 200)
+    y = np.sin(x) + noise(x.shape[0])
+    test = qutip.Cubic_Spline(x[0], x[-1], y)
+    expected = scipy.interpolate.CubicSpline(x, y, bc_type="natural")
+    # We use the bc_type="natural", i.e. zero second-derivatives at the
+    # boundaries because that matches qutip.
+    np.testing.assert_allclose(test(x), expected(x), atol=1e-10)
+    centres = 0.5 * (x[:-1] + x[1:])
+    # Test some points not in the original.
+    for point in centres:
+        assert np.abs(test(point) - expected(point)) < 1e-10, "Scalar argument"
+
+
+# We make a bunch of different test cases for the coefficients, all as a
+# three-tuple of
+#   (string, other, type)
+# where `string` is the string-format time-dependence (we use this as the
+# reference), `other` is what will become the time-dependence for this part of
+# the test, and `type` is a string in ["spline", "function", "string"] which
+# tells us how to interpret the type of `other`.  If `type` is "function" or
+# "string", then nothing is done to `other` and it is used as-is.  If `type` is
+# "spline", then `other` is given as a vectorised function which we turn into a
+# qutip.Cubic_Spline using appropriate `x` and `y` values.
+_case_real_spline = ('0.25*sin(t)', lambda t: 0.25*np.sin(t), 'spline')
```

Comment:
The reference could be converted to a function to save time.

Reply:
Sorry, I'm not sure what you mean.

The reason there's a layer of indirection at all is so that the test case can "feed in" the number of times that are to be used, but I think I had to add in a special case for splines anyway, so I might re-think that whole parametrisation.

Reply:
What I meant was to use a function format coefficient instead of a string-format time-dependence for the reference. String format need compilation which take about a sec each time.

Reply:
Ah, so the original reason that the base was a string was to accommodate `brmesolve`, which can't accept the function forms.  I suspect they were originally just copy+paste code, so they were all the same.

Are you planning a complete overhaul of `brmesolve`?  If so, this switch-over could wait until then.  If not, I'll think rather more about how to handle these cases so it doesn't turn into a huge mess.